File: nmLabReport/lab21.c
#include <stdio.h>

#define MAX 10

// Function to calculate factorial
int factorial(int n) {
    int fact = 1;
    for (int i = 2; i <= n; i++) {
        fact *= i;
    }
    return fact;
}

// Function to calculate the backward difference table
void backwardDifferenceTable(float x[], float y[][MAX], int n) {
    for (int i = 1; i < n; i++) {
        for (int j = n - 1; j >= i; j--) {
            y[j][i] = y[j][i - 1] - y[j - 1][i - 1];
        }
    }
}

// Function to calculate the derivative using Newton's Backward Difference
float newtonBackwardDifference(float x[], float y[][MAX], int n, float value) {
    float h = x[1] - x[0];
    float u = (value - x[n - 1]) / h;
    float derivative = 0;

    for (int i = 1; i < n; i++) {
        derivative += (y[n - 1][i] * i * pow(u, i - 1)) / factorial(i);
    }

    return derivative / h;
}

int main() {
    int n;
    float x[MAX], y[MAX][MAX], value;

    // Input number of data points
    printf("Enter the number of data points: ");
    scanf("%d", &n);

    // Input data points
    printf("Enter the data points (x and y):\n");
    for (int i = 0; i < n; i++) {
        printf("x[%d] = ", i);
        scanf("%f", &x[i]);
        printf("y[%d] = ", i);
        scanf("%f", &y[i][0]);
    }

    // Input the value to find the derivative at
    printf("Enter the value to find the derivative at: ");
    scanf("%f", &value);

    // Calculate the backward difference table
    backwardDifferenceTable(x, y, n);

    // Calculate the derivative using Newton's Backward Difference
    float result = newtonBackwardDifference(x, y, n, value);

    // Output the result
    printf("The derivative at x = %.2f is %.4f\n", value, result);

    return 0;
}

File: nmLabReport/lab21.py
import math

def factorial(n):
    fact = 1
    for i in range(2, n + 1):
        fact *= i
    return fact

def backward_difference_table(x, y, n):
    for i in range(1, n):
        for j in range(n - 1, i - 1, -1):
            y[j][i] = y[j][i - 1] - y[j - 1][i - 1]

def newton_backward_difference(x, y, n, value):
    h = x[1] - x[0]
    u = (value - x[n - 1]) / h
    derivative = 0

    for i in range(1, n):
        derivative += (y[n - 1][i] * i * math.pow(u, i - 1)) / factorial(i)

    return derivative / h

def main():
    n = int(input("Enter the number of data points: "))
    x = []
    y = [[0 for _ in range(n)] for _ in range(n)]

    print("Enter the data points (x and y):")
    for i in range(n):
        x.append(float(input(f"x[{i}] = ")))
        y[i][0] = float(input(f"y[{i}] = "))

    value = float(input("Enter the value to find the derivative at: "))

    backward_difference_table(x, y, n)
    result = newton_backward_difference(x, y, n, value)

    print(f"The derivative at x = {value:.2f} is {result:.4f}")

if __name__ == "__main__":
    main()